whylog/teacher/__init__.py

```diff
@@ -156,53 +173,41 @@ def set_causes_relation(self, relation):
         """
         pass
 
-    def _validate_primary_keys(self):
-        errors = []
-        for line_id in six.iterkeys(self._parsers):
-            primary_key = self._parsers[line_id].primary_key
-            pattern_match = self.pattern_assistant.get_pattern_match(line_id)
-            group_numbers = pattern_match.param_groups.keys()
-            if set(primary_key) - set(group_numbers):
-                errors.append(InvalidPrimaryKeyProblem(line_id, primary_key, group_numbers))
-        return ValidationResult(errors=errors, warnings=[])
-
-    def _validate_pattern_names(self):
-        errors = []
+    def _rule_problems(self):
+        problems = []
+        if self.effect_id is None:
+            problems.append(NoEffectParserProblem())
+        if len(self._parsers) < 2:
+            problems.append(OneParserRuleProblem())
+        return problems
+
+    def _parser_problems(self):
+        problems = {}
         names_blacklist = self._get_names_blacklist()
-        for line_id in six.iterkeys(self._parsers):
-            parser = self._parsers[line_id]
-            blacklist_except_name = set(names_blacklist) - set([parser.name])
-            if names_blacklist.count(parser.name) > 1 or \
-                    not self.config.is_free_parser_name(parser.name, blacklist_except_name):
-                errors.append(NotUniqueParserNameProblem(line_id))
-        return ValidationResult(errors=errors, warnings=[])
-
-    def _validate_log_type(self):
-        errors = []
-        for line_id in six.iterkeys(self._parsers):
-            parser = self._parsers[line_id]
-            if parser.log_type is None:
-                errors.append(NotSetLogTypeProblem(line_id))
-        return ValidationResult(errors=errors, warnings=[])
-
-    def _validate_constraints(self):
-        return ValidationResult.result_from_results(
-            [constraint.validate() for constraint in six.itervalues(self._constraint_base)]
-        )
+        for parser_id in six.iterkeys(self._parsers):
+            parser = self._parsers[parser_id]
+            parser_problems = []
+            parser_problems += parser.validate_name(self.config, names_blacklist)
+            parser_problems += parser.validate_log_type()
+            parser_problems += self.pattern_assistant.validate(parser_id)
+            problems[parser_id] = parser_problems
+        return problems
+
+    def _constraint_problems(self):
+        problems = {}
+        for constraint_id in six.iterkeys(self._constraint_base):
+            constraint = self._constraint_base[constraint_id]
```

Comment:
this and above line reimplements `six.iteritems`. This whole function should be a list comprehension.